Prompt API: Add overview for Prompt API (#43848)
mdn/content: files/en-us/web/api/languagemodel/clone/index.md

---
title: "LanguageModel: clone() method"
short-title: clone()
slug: Web/API/LanguageModel/clone
page-type: web-api-instance-method
spec-urls: https://webmachinelearning.github.io/prompt-api/
---

{{APIRef("Prompt API")}}{{SecureContext_Header}}

The **`clone()`** method of the {{domxref("LanguageModel")}} interface creates a copy of the current session, including its full context window state. The cloned session can be used independently without affecting the original.

Because both the original and the clone share the same context history up to the point of cloning, you can explore multiple response paths or test variations without starting from scratch.

For example, you might build a shared context using {{domxref("LanguageModel.append()", "append()")}} or early {{domxref("LanguageModel.prompt()", "prompt()")}} `prompt()` calls, clone the session, and then send different follow-up prompts to each clone in parallel.

## Syntax

```js-nolint
clone()
clone(options)
```

### Parameters

- `options` {{optional_inline}}

>   - : Represents the options that can be passed. If this argument is absent, the `options` from the original session are used (not copied). Options include:

Reviewer: So the clone would be cancelled by the same signal as the original.? Maybe

```suggestion
  - : Represents the options that can be passed. If this argument is absent, the `options` from the original session, such as its abort signal, are used.
    Options include:
```

    - `signal` — An {{domxref("AbortSignal")}} to cancel the clone operation.

### Return value

A {{jsxref("Promise")}} that resolves with a new {{domxref("LanguageModel")}} instance. The new instance has the same internal model state, context history, and configuration as the original session at the time of cloning.

### Exceptions

- `OperationError` {{domxref("DOMException")}}
  - : Thrown if the clone operation fails.
- `AbortError` {{domxref("DOMException")}}
  - : Thrown if the operation was cancelled via the `signal` option.

## Examples

> ### Exploring multiple response paths

Reviewer: Normally we prefer more words in examples. The main reason here is that you explain the code, creating links to bits that might interest people. 

Not sure strictly necesssary in this case because the code is pretty clear and self explanatory.

Reviewer: Just to be a bit clear. It is necessary, but you have the comments inline in the code. 

I'd certainly do more, but perhaps you think it unnecessary. What do you think a high schooler would need to use and understand this broadly in this context?

Author: Next commit:

The following example shows how to explore different response paths. First, it creates a single session with the start of a story. Then it clones the original session twice before prompting for different endings. This approach preserves the original session in case more exploration is wanted.

Author: This one has text also.

```js

> const session = await LanguageModel.create({

Reviewer: ```suggestion
// Set up system and user context
const session = await LanguageModel.create({
```

spec-urls: https://webmachinelearning.github.io/prompt-api/
---

{{APIRef("Prompt API")}}{{SecureContext_Header}}

The **`clone()`** method of the {{domxref("LanguageModel")}} interface creates a copy of the current session, including its full context window state. The cloned session can be used independently without affecting the original.

Because both the original and the clone share the same context history up to the point of cloning, you can explore multiple response paths or test variations without starting from scratch.

For example, you might build a shared context using {{domxref("LanguageModel.append()", "append()")}} or early {{domxref("LanguageModel.prompt()", "prompt()")}} `prompt()` calls, clone the session, and then send different follow-up prompts to each clone in parallel.

## Syntax

```js-nolint
clone()
clone(options)
```

### Parameters

- `options` {{optional_inline}}
  - : Represents the options that can be passed. If this argument is absent, the `options` from the original session, such as its abort signal, are used.
    Options include:
    - `signal` — An {{domxref("AbortSignal")}} to cancel the clone operation.

### Return value

> A {{jsxref("Promise")}} that resolves with a new {{domxref("LanguageModel")}} instance. The new instance has the same internal model state, context history, and configuration as the original session at the time of cloning.

Reviewer: I think we already say part of this above.

```suggestion
A {{jsxref("Promise")}} that resolves with a cloned {{domxref("LanguageModel")}} instance.
```

### Exceptions

- `OperationError` {{domxref("DOMException")}}
  - : Thrown if cloning fails for any other reason not listed in the other exception types.
- `AbortError` {{domxref("DOMException")}}

>   - : Thrown if the operation was cancelled via the `signal` option.

Reviewer: Alphabetic by default. 

```suggestion
- `AbortError` {{domxref("DOMException")}}
  - : Thrown if the operation was cancelled via the `signal` option.
- `OperationError` {{domxref("DOMException")}}
  - : Thrown if cloning fails for any other reason not listed in the other exception types.
```

## Examples

### Exploring multiple response paths

The following example shows how to explore different response paths. First, it creates a single session with the start of a story. Then it clones the original session twice before prompting for different endings. This approach preserves the original session in case more exploration is wanted.

```js
const session = await LanguageModel.create({
  initialPrompts: [
    { role: "system", content: "You are a creative writing assistant." },
  ],
});

await session.append(
  "The story begins in a small coastal town during a storm.",
);

const [clone1, clone2] = await Promise.all([session.clone(), session.clone()]);

const [ending1, ending2] = await Promise.all([
  clone1.prompt("Write a happy ending."),
  clone2.prompt("Write a mysterious ending."),
]);

console.log("Happy ending:", ending1);
console.log("Mysterious ending:", ending2);
```

### Cloning to retry after a context overflow

> 

Reviewer: When I ask for description in one case, please do it everywhere. Or argue that it completely unnecessary for given reasons. As it is now, I'm looking at this in despair.

- `AbortError` {{domxref("DOMException")}}
  - : Thrown if the operation was cancelled via the `signal` option.
- `OperationError` {{domxref("DOMException")}}
  - : Thrown if cloning fails for any other reason not listed in the other exception types.

## Examples

### Exploring multiple response paths

The following example shows how to explore different response paths. First, it creates a single session with the start of a story. Then it clones the original session twice before prompting for different endings. This approach preserves the original session in case more exploration is wanted.

```js
const session = await LanguageModel.create({
  initialPrompts: [
    { role: "system", content: "You are a creative writing assistant." },
  ],
});

await session.append(
  "The story begins in a small coastal town during a storm.",
);

const [clone1, clone2] = await Promise.all([session.clone(), session.clone()]);

const [ending1, ending2] = await Promise.all([
  clone1.prompt("Write a happy ending."),
  clone2.prompt("Write a mysterious ending."),
]);

console.log("Happy ending:", ending1);
console.log("Mysterious ending:", ending2);
```

### Cloning to retry after a context overflow

> This example uses a checkpoint and rollback pattern to save the state of a session before attempting to append a large amount of data. Cloning the session before calling `append()` ensures that the app has a way to restore to a safe state before attempting something risky.

Reviewer: Perhaps 

```suggestion
This example uses a checkpoint and rollback pattern to save the state of a session before attempting to append a large amount of data. Cloning the session before calling `append()` allows the app to restore the state if the context window is exceeded.
```

```js
const veryLargeDocument = "This is my very long story...";
let session = await LanguageModel.create();
const checkpoint = await session.clone();

try {
  await session.append(veryLargeDocument);
} catch (err) {
  if (err.name === "QuotaExceededError") {
    console.warn("Document too large.");
    session = checkpoint;
  }

> }

Reviewer: Ah, I guess this is why you use the term throw - all the promise code is wrapped.
